leapp/reporting/__init__.py

```diff
@@ -5,102 +5,254 @@
 
-    audience = fields.List(fields.StringEnum(choices=['developer', 'sysadmin']))
-    """
-    Who is the main audience of this report entry
-    """
+    def apply(self, report):
+        """Add report entry to the report dict based on provided path"""
+        _add_to_dict(report, self.path, self.value, self.is_list)
 
-    flags = fields.List(fields.StringEnum(choices=['inhibitor']))
-    """
-    Flags can give a special meaning to each report entry, currently supported flags:
 
-        inhibitor - This report entry will inhibit the upgrade process
-    """
+class Title(BasePrimitive):
+    """Report title"""
+    @property
+    def path(self):
+        return ('title',)
+
+
+class Summary(BasePrimitive):
+    """Report summary"""
+    @property
+    def path(self):
+        return ('summary',)
+
+
+class Severity(BasePrimitive):
+    """Report severity"""
+
+    INFO = 'info'
+    LOW = 'low'
+    MEDIUM = 'medium'
+    HIGH = 'high'
+
+    def __init__(self, value=None):
+        severities = (self.INFO, self.LOW, self.MEDIUM, self.HIGH)
+        if value not in severities:
+            raise ValueError('Severity should be one of: "{}"'.format(' '.join(severities)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('severity',)
+
+
+class Audience(BasePrimitive):
+    """Target audience of the report"""
+    def __init__(self, value=None):
+        if not isinstance(value, str):
+            raise TypeError('Value of "{}" must be a string'.format(self.__class__.__name__))
+        audiences = ('sysadmin', 'developer')
+        if value not in audiences:
+            raise ValueError('Audience should be one of: "{}"'.format(' '.join(audiences)))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('audience',)
+
+
+class Flags(BasePrimitive):
+    """Report flags"""
 
+    INHIBITOR = 'inhibitor'
+    FAILURE = 'failure'
 
-def report(title=None, detail=None, renderers=None, audience=None, flags=None, severity=None):
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
+        self._value = value
+
+    @property
+    def path(self):
+        return ('flags',)
+
+
+class Tags(BasePrimitive):
+    """Report tags"""
+
+    ACCESSIBILITY = 'accessibility'
+    AUTHENTICATION = 'authentication'
+    COMMUNICATION = 'communication'
+    DRIVERS = 'drivers'
+    EMAIL = 'email'
+    FILESYSTEMS = 'filesystems'
+    HIGH_AVAILABILITY = 'high availability'
+    NETWORK = 'network'
+    SANITY = 'sanity'
+    SECURITY = 'security'
+    SELINUX = 'selinux'
+    SERVICES = 'services'
+    TIME_MANAGEMENT = 'time management'
+    TOOLS = 'tools'
+    UPGRADE_PROCESS = 'upgrade process'
+
+    def __init__(self, value=None):
+        if not isinstance(value, list):
+            raise TypeError('Value of "{}" must be a list'.format(self.__class__.__name__))
```

Comment:
unnecessary format

Reply:
I left the format only in the base class...